concordium-std/src/test_infrastructure/trie.rs

```diff
@@ -100,20 +81,77 @@ impl StateTrie {
         }
         key
     }
+    /// Returns true if the subtree corresponding to the given key is
+    /// already locked by an existing iterator, false otherwise.
+    fn is_locked(&self, prefix: &[usize]) -> bool {
+        self.iterators.borrow().keys().any(|p| {
+            let shortest = min(p.len(), prefix.len());
+            p[..shortest] == prefix[..shortest]
+        })
+    }
+
+    pub fn create_entry(&mut self, key: &[u8]) -> Result<StateEntryTest, ContractStateError> {
+        let indexes = Self::to_indexes(key);
+        if self.is_locked(&indexes) {
+            return Err(ContractStateError::SubtreeLocked);
+        }
+        let data = self.nodes.create(&indexes);
+        let entry = self.construct_state_entry_test(indexes, data, key.to_vec());
+        Ok(entry)
+    }
+
+    pub fn lookup(&self, key: &[u8]) -> Option<StateEntryTest> {
+        let indexes = Self::to_indexes(key);
+        self
+            .nodes
+            .lookup(&indexes)
+            .map(|data| self.construct_state_entry_test(indexes, data, key.to_vec()))
+    }
+
+    pub fn delete_entry(&mut self, entry: StateEntryTest) -> Result<(), ContractStateError> {
+        let indexes = Self::to_indexes(&entry.get_key().expect("key must exist"));
+        if self.is_locked(&indexes) {
+            return Err(ContractStateError::SubtreeLocked);
+        }
+        match self.entry_map.borrow_mut().remove(&entry.state_entry_id) {
+            Some(indexes) => self.nodes.delete_data(&indexes),
+            None => Err(ContractStateError::EntryNotFound), /* Entry did not exist. Only happens when entry was deleted using
+                            * delete_prefix. */
+        }
+    }
+
+    pub fn iterator(&self, prefix: &[u8]) -> Result<Iter, ContractStateError> {
+        let index_prefix = StateTrie::to_indexes(prefix);
+        if let Some(iter) = self.iterators.borrow().get(&index_prefix) {
+            if iter.clone_count() <= u16::MAX as usize {
+                return Ok(iter.clone());
+            }
+            return Err(ContractStateError::IteratorLimitForPrefixExceeded);
+        }
+        let node = self.nodes.lookup_node(&index_prefix).ok_or(ContractStateError::EntryNotFound)?;
+        let iter = Iter::new(self, index_prefix.clone(), node);
+        self.iterators.borrow_mut().insert(index_prefix, iter.clone());
+        Ok(iter)
+    }
+
+    pub fn delete_iterator(&mut self, iterator: Iter) -> bool {
+        if iterator.clone_count() > 2 {
+            println!("clone count = {}", iterator.clone_count());
+            return true;
+        }
+        println!("2");
+        let initial_count = self.iterators.borrow().len();
+        self.iterators.borrow_mut().retain(|_, iter| iter != &iterator);
```

Comment:
An optimization here would be that the iterator keeps track of its own prefix, so we can use `remove(iterator.key)` instead. But the current solution is probably fine for testing.